### PR TITLE
ddl: fix prepare stmt with partition related ddl stmts

### LINKED ISSUE #54283: runtime error: slice bounds out of range when exec prepared statement with leading newline
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Run below sql 
```
create database if not exists test;
use test;
drop table if exists aaa;
--- test case A
prepare stmt from "\ncreate table aaa(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000) maxvalue partition";
execute stmt;

use test; drop table if exists bbb;
create table bbb(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000);
-- test case B
prepare stmt2 from "\nalter table bbb last partition less than (20000)";
execute stmt2;

-- test case C
prepare stmt3 from "\nalter table bbb first partition less than (200)";
execute stmt3;
```

### 2. What did you expect to see? (Required)
Prepared statement with leading newline to create partition table should succeed.

### 3. What did you see instead (Required)
```
mysql> create database if not exists test;
Query OK, 0 rows affected, 1 warning (0.04 sec)

use test;
drop table if exists aaa;
--- test case A
prepare stmt from "\ncreate table aaa(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000) maxvalue partition";
execute stmt;mysql> use test;
Database changed
drop table if exists aaa;
--- test case A
prepare stmt from "\ncreate table aaa(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000) maxvalue partition";
execute stmt;mysql> drop table if exists aaa;
Query OK, 0 rows affected, 1 warning (0.04 sec)

--- test case A
prepare stmt from "\ncreate table aaa(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000) maxvalue partition";
execute stmt;mysql> --- test case A
    -> prepare stmt from "\ncreate table aaa(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000) maxvalue partition";
execute stmt;prepare stmt from "\ncreate table aaa(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000) maxvalue partition";
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 1 near "--- test case A
prepare stmt from "\ncreate table aaa(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000) maxvalue partition"" execute stmt;
mysql> execute stmt;
ERROR 1105 (HY000): runtime error: slice bounds out of range [163:162]
mysql> use test; drop table if exists bbb;
Database changed
create table bbb(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000);
-- test case B
prepare stmt2 from "\nalter table bbb last partition less than (20000)";
execute stmt2;
Query OK, 0 rows affected (0.09 sec)

mysql> create table bbb(id bigint primary key) partition by range (id) interval (100) first partition less than (100) last partition less than (10000);
Query OK, 0 rows affected (0.08 sec)

-- test case B
prepare stmt2 from "\nalter table bbb last partition less than (20000)";
execute stmt2;
mysql> -- test case B
Query OK, 0 rows affected (0.03 sec)
prepare stmt2 from "\nalter table bbb last partition less than (20000)";
execute stmt2;

mysql> prepare stmt2 from "\nalter table bbb last partition less than (20000)";
Query OK, 0 rows affected, 1 warning (0.04 sec)
execute stmt2;

mysql> execute stmt2;
ERROR 1105 (HY000): runtime error: slice bounds out of range [49:48]

mysql> -- test case C
Query OK, 0 rows affected, 2 warnings (0.04 sec)

prepare stmt3 from "\nalter table bbb first partition less than (200)";
execute stmt3;mysql> prepare stmt3 from "\nalter table bbb first partition less than (200)";
Query OK, 0 rows affected, 1 warning (0.04 sec)
execute stmt3;
mysql> execute stmt3;
ERROR 1105 (HY000): runtime error: slice bounds out of range [48:47]

```

### 4. What is your TiDB version? (Required)

Release Version: v8.2.0-alpha
Edition: Community
Git Commit Hash: 1469fcfd5914b165cef5a026dcc05d8e01c4882f
Git Branch: heads/refs/tags/v8.2.0-alpha
UTC Build Time: 2024-06-26 11:46:38
GoVersion: go1.21.10
Race Enabled: false
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #54283

Problem Summary: We use `OriginTextPosition` to split prepared ddl stmts. But it is wrong when query has prefixed blank, like '\ncreate ...'. So we should call `strings.index` to regenerate it.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/ddl_api.go [hunk 1]
```diff
@@ -4641,7 +4641,12 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
 			AppendPartitionDefs(partInfo, &buf, sqlMode)
 
 			syntacticSugar := spec.Partition.PartitionMethod.OriginalText()
-			syntacticStart := spec.Partition.PartitionMethod.OriginTextPosition()
+			syntacticStart := strings.Index(query, syntacticSugar)
+			if syntacticStart == -1 {
+				logutil.DDLLogger().Error("Can't find related PARTITION definition in prepare stmt",
+					zap.String("PARTITION defiition", syntacticSugar), zap.String("prepare stmt", query))
```
Comment (mjonss) [inline pkg/ddl/ddl_api.go:4647]:
```suggestion
					zap.String("PARTITION definition", syntacticSugar), zap.String("prepare stmt", query))
```

File: pkg/ddl/partition.go [hunk 2]
```diff
@@ -615,7 +615,12 @@ func buildTablePartitionInfo(ctx sessionctx.Context, s *ast.PartitionOptions, tb
 			AppendPartitionDefs(tbInfo.Partition, &buf, sqlMode)
 
 			syntacticSugar := s.Interval.OriginalText()
-			syntacticStart := s.Interval.OriginTextPosition()
+			syntacticStart := strings.Index(query, syntacticSugar)
+			if syntacticStart == -1 {
+				logutil.DDLLogger().Error("Can't find INTERVAL definition in prepare stmt",
+					zap.String("INTERVAL defiition", syntacticSugar), zap.String("prepare stmt", query))
```
Comment (mjonss) [inline pkg/ddl/partition.go:621]:
```suggestion
					zap.String("INTERVAL definition", syntacticSugar), zap.String("prepare stmt", query))
```

**Review by mjonss — COMMENTED** [review]
Would it be possible to fix this by correcting the `SetOriginTextPosition`? My concern is that we now use OriginTextPosition less, and should try to either replace OriginTextPosition completely OR try to fix the current use case as well.
What do you think, is it possible to:
- fix by SetOriginTextPosition
- replace all use of OriginTextPosition (if they can be wrong in case of prepared statement)
- Only fix these broken ones and keep a mix of `strings.Index()` and `OriginTextPosition`?

**Comment by Defined2014** [comment]
> Would it be possible to fix this by correcting the `SetOriginTextPosition`? My concern is that we now use OriginTextPosition less, and should try to either replace OriginTextPosition completely OR try to fix the current use case as well. What do you think, is it possible to:
> 
> * fix by SetOriginTextPosition
> * replace all use of OriginTextPosition (if they can be wrong in case of prepared statement)
> * Only fix these broken ones and keep a mix of `strings.Index()` and `OriginTextPosition`?

* For `OriginTextPosition`, I think it is right. Because in origin text, it has a `\n` at the beginning, prepare stmt will remove it by strings.Trim().
* Apart from these three places, `OriginTextPosition` is not used anywhere else.

Maybe we can leave spaces in the prefix and suffix after prepared stmt to fix this problem(not sure does it affects other problems). What do you think? @mjonss

**Review by mjonss — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Review by yudongusa — APPROVED** [review]

**Pull request merged**